Aliasing

WuTheFWasThat opened the pull request.

za3k commented:
Jeff, I thought allAncestors might help out if you wanted to make marks functionality fast again. You may need to figure out the delta-ancestry on attach/detach/move.

WuTheFWasThat commented:
isn't allAncestors O(subtree size) instead of O(depth) though?

za3k commented:
No, allAncestors is O(ancestors). It's performing a depth-first search with marking, through the ancestry only. Or if it's not I wrote it wrong, I really have no idea why I committed that without testing it. My bad.

WuTheFWasThat commented:
oh, yeah sorry. i was just thinking it reversed in my head, oops!  that function seems useful

za3k commented:
Ready to merge.
All tree operations are O(ancestry) at most now (without clones that's O(depth), will be something similar for clones)

Fixed: Marks, movement, cloning were all O(subtree) at least previously

WuTheFWasThat commented:
still not convinced of correctness.  issue is when the same clone appears multiple times in a subtree, removing the first one will remove incorrectly from the ancestor right?  if true, write a test case that is broken, before fixing

WuTheFWasThat commented:
nvm, i think i misread the code.  test case would still be good though!

WuTheFWasThat commented:
instead of delta ancestry would it be simpler to do counts?  i'm okay either way though, as long as it works

WuTheFWasThat commented on assets/js/data.coffee:
```diff
@@ -249,11 +263,65 @@ class Data
+    parentCount = 0
+    for parent_id in (@store.getParents id)
+      parentCount += @countInstances parent_id # Always exactly once under every parent
+    return parentCount
```
this is a sort of expensive thing to do, and it's only used for exactlyOneInstance, right?  i think exactlyOneInstance could just check # of parents

za3k replied:
1. exactlyOneInstance will return the incorrect result if it literally only checks the number of parents, although it could be special-cased, yes.
2. this isn't really a very expensive operation, could you be more specific as to how you think it will take a long time? if we added memoization to the parent count, it would be O(ancestors) which is optimal for exactlyOneInstance. If we don't it double-counts some diamond-shaped ancestry but it's still quite fast.

za3k replied:
Fined-tuned countInstances. exactlyOneInstance now has optimal complexity without maintaining extra data. (you could argue this is a little more complicated to read, but I think

WuTheFWasThat commented on assets/js/data.coffee:
```diff
@@ -249,11 +263,65 @@ class Data
+    visit id
+    ancestors
+
+  # Precondition: row.id == id
```
perhaps just pass in row?

za3k replied:
Because it's a "diff" style operation I thought this was less confusing.

WuTheFWasThat commented on assets/js/data.coffee:
```diff
@@ -384,16 +460,25 @@ class Data
+    # Precondition: tree is not already circular
+    # Rather than checking for each descendent, whether that descendent is an ancestor (old method)
+    # Instead, check if the row is an ancestor of the new parent, which is sufficient.
+    _.contains (@allAncestors parent.id, { inclusive: true }), row.id
```
i think we need to get the set of clones underneath the row, as well, right?

za3k replied:
Nope.

za3k replied:
To explain further, if there WAS a clone underneath the row which was an ancestor of 'parent', then 'row' would also be an ancestor of 'parent'

WuTheFWasThat replied:
ah, you're right.  cool!

WuTheFWasThat commented on test/tests/clone.coffee:
```diff
@@ -0,0 +1,299 @@
+    console.log '\nDEBUG STEP 3'
+    console.log 'marks under 1', (t.store.getMarks 1)
+    console.log 'marks under 3', (t.store.getMarks 3)
+    console.log 'all marks', (do t.store.getAllMarks)
```
these log statements can go, heh

WuTheFWasThat commented on assets/html/layout.jade:
```diff
@@ -27,6 +27,7 @@ html
     script(src='/jquery.cookie.min.js')
     script(src='/lodash.js')
     script(src='/virtual-dom.js')
+    script(src='/bimap.js')
```
bimap's not being used anymore, I think?  let's remove it if so

za3k commented:
Not sure how to do counts to get the correct answer, but I think deltaAncestry isn't more than a factor of two more expensive than counts would be. (except maybe that _.difference is implemented a really stupid way)

za3k commented:
Can you add any test cases when you review this weekend?

WuTheFWasThat commented on assets/js/data.coffee:
```diff
@@ -249,11 +263,67 @@ class Data
+    rowAncestry = @allAncestors parentId, { inclusive: options.inclusive }
+    idAncestry = []
+    for otherParentId in _.without (@getParents row), parentId
+      idAncestry = _.union idAncestry, (@allAncestors otherParentId, { inclusive: true })
```
instead of this union, could you just do `@allAncestors id, {inclusive: false}` ?

za3k replied:
Nope. We're trying to figure which things are ancestors of id only via one parent, and not via any of the other parents--taking "via all parents" and subtracting "via this parent" does not work.

WuTheFWasThat replied:
oh i see. it would work if you compared the ancestries before and after, i guess?

WuTheFWasThat commented on assets/js/data.coffee:
```diff
@@ -268,20 +338,26 @@ class Data
   attachChild: (row, child, index = -1) ->
-    children = @attachChildren row, [child], index
+    @attachChildren row, [child], index
+    return child
```
i think there's no need to return child here

za3k replied:
It gets used in one of the mutators, I believe.

za3k replied:
Nope, it doesn't, but I still think it's fine to return--less confusing than returning [child], which is the default otherwise.

WuTheFWasThat commented on assets/js/data.coffee:
```diff
@@ -324,7 +403,6 @@ class Data
     firstDifference = commonAncestry.length
     return [common, ancestors1[firstDifference..], ancestors2[firstDifference..]]
 
-  # extends a row's path using descendents (used when moving blocks around)
```
maybe restore this comment?  it's a sort of weird function, after all

za3k replied:
Oops, not sure how this got deleted, that wasn't on purpose

WuTheFWasThat commented on assets/js/data.coffee:
```diff
@@ -158,63 +161,71 @@ class Data
       return false
 
-    oldmark = @getMark row
+    oldmark = (@store.getMarks id)[id]
```
this can still be `getMark`, i think?

za3k replied:
Trying to remove a failure case around: or '', but you could be right.

za3k replied:
Oh no I see, it's because 'row' no longer gets passed in.

WuTheFWasThat commented:
deltaAncestry is fine.  I'm not worried about its complexity

I added some test cases, caught a pair of bugs.  i fixed the first one - i gotta sleep though so you'll have to fix the other

za3k commented on test/tests/clone.coffee:
```diff
@@ -0,0 +1,538 @@
+        ] }
+        { text: 'blah', children: [
+          'blah'
+          { text: 'Not a clone', children: [
```
If you're going to clone something, don't call it "Not a clone".

za3k commented on test/tests/clone.coffee:
```diff
@@ -0,0 +1,538 @@
+        ] }
+      ] }
+    ]
+    t.expectMarks {}
```
This test case is failing because it is incorrect and the code is correct.

There should be one mark here (see "deletes marks only on last clone delete")

Marks are a per-id feature, not a per-row feature (nothing is a per-row feature, currently)

WuTheFWasThat replied:
are you sure it's wrong?  I understand marks are a per-id feature.  but all instances of that id disappeared, right?  that final t.expect passes but the expectMarks doesn't, which at the very least seems inconsistent

sorry for the bad names for the rows

za3k replied:
Oh oops, didn't read carefully enough. Yes, I see the problem. Will fix it Monday.

za3k commented:
Tests pass again, merged with master
